cmd/fleetctl/apple_mdm.go

```diff
@@ -44,124 +52,137 @@ func appleMDMCommand() *cli.Command {
+		},
 		Subcommands: []*cli.Command{
-			appleMDMSetupSCEPCommand(),
-			appleMDMSetupAPNSCommand(),
-			appleMDMSetupDEPCommand(),
+			generateMDMAppleCommand(),
 		},
 	}
 }
 
-func appleMDMSetupSCEPCommand() *cli.Command {
-	// TODO(lucas): Define workflow when SCEP CA certificate expires.
-	var (
-		validityYears      int
-		cn                 string
-		organization       string
-		organizationalUnit string
-		country            string
-	)
+func generateMDMAppleCommand() *cli.Command {
 	return &cli.Command{
-		Name:  "scep",
-		Usage: "Create SCEP certificate authority",
+		Name:  "mdm-apple",
+		Usage: "Generates certificate signing request (CSR) and key for Apple Push Notification Service (APNs) and certificate and key for Simple Certificate Enrollment Protocol (SCEP) to turn on MDM features.",
 		Flags: []cli.Flag{
-			&cli.IntFlag{
-				Name:        "validity-years",
-				Usage:       "Validity of the SCEP CA certificate in years",
-				Required:    true,
-				Destination: &validityYears,
+			&cli.StringFlag{
+				Name:     "email",
+				Usage:    "The email address to send the signed APNS csr to.",
+				Required: true,
 			},
 			&cli.StringFlag{
-				Name:        "cn",
-				Usage:       "Common name to set in the SCEP CA certificate",
-				Required:    true,
-				Destination: &cn,
+				Name:     "org",
+				Usage:    "The organization requesting the signed APNS csr.",
+				Required: true,
 			},
 			&cli.StringFlag{
-				Name:        "organization",
-				Usage:       "Organization to set in the SCEP CA certificate",
-				Required:    true,
-				Destination: &organization,
+				Name:     "apns-key",
+				Usage:    "The output path for the APNs private key.",
+				Value:    apnsKeyPath,
+				Required: true,
 			},
 			&cli.StringFlag{
-				Name:        "organizational-unit",
-				Usage:       "Organizational unit to set in the SCEP CA certificate",
-				Required:    true,
-				Destination: &organizationalUnit,
+				Name:     "scep-cert",
+				Usage:    "The output path for the SCEP CA certificate.",
+				Value:    scepCACertPath,
+				Required: true,
 			},
 			&cli.StringFlag{
-				Name:        "country",
-				Usage:       "Country to set in the SCEP CA certificate",
-				Required:    true,
-				Destination: &country,
+				Name:     "scep-key",
+				Usage:    "The output path for the SCEP CA private key.",
+				Value:    scepCAKeyPath,
+				Required: true,
 			},
 		},
 		Action: func(c *cli.Context) error {
-			certPEM, keyPEM, err := scep_ca.Create(validityYears, cn, organization, organizationalUnit, country)
+			email := c.String("email")
+			org := c.String("org")
+			apnsKeyPath := c.String("apns-key")
+			scepCACertPath := c.String("scep-cert")
+			scepCAKeyPath := c.String("scep-key")
+
+			fmt.Fprintf(
+				c.App.Writer,
+				`Sending certificate signing request (CSR) for Apple Push Notification service (APNs) to %s...
+Generating APNs key, Simple Certificate Enrollment Protocol (SCEP) certificate, and SCEP key...
```

Comment:
should we print each of these lines separately as the processes are completed? To clarify, I mean something like:

```go
fmt.Fprintf(c.App.Writer, "Sending certificate signing request (CSR) for Apple Push Notification service (APNs) to %s..., email)

// create apns csr and send to fleetdm.com
// ....

fmt.Fprintf(c.App.Writer, "Generating APNs key, Simple Certificate Enrollment Protocol (SCEP) certificate, and SCEP key...")

// init scep ca
// ...
```

This way it's more clear in the output if there was an error during the APNS Key generation.

Reply:
We cannot send the apns csr before generating the apns key. The key must be generated first. Therefore, we can't print these messages in the same order.

@noahtalerman Should I rearrange the messages so that they are printed in the same order. Is it okay if we exit early with an error message if we encounter an error before generating the scep cert and key?

Reply:
> Should I rearrange the messages so that they are printed in the same order. 

@michalnicp @roperzh I think the plan was to print the lines together. This may not have been clear in this specs. Can we do this? 

This way, if anything goes wrong (email sending / certificate or key generation) the user sees an error that instructs them to try again.

> Is it okay if we exit early with an error message if we encounter an error before generating the scep cert and key?

The plan is to print all the success messages printed together as one message. This also might have not been clear in the specs. I think this means that we don't need to exit early. The user either sees all success messages or an error.

Reply:
in that case seems like we're good to go? please disregard my original comment.

Reply:
@noahtalerman I am not clear on what is considered "success messages". Is it this

```
Sending certificate signing request (CSR) for Apple Push Notification service (APNs) to {mo@acme.com}... 
Generating APNs key, Simple Certificate Enrollment Protocol (SCEP) certificate, and SCEP key...

Success! 
```

or this

```
Success! 
```

> The user either sees all success messages or an error.

In the example error given in https://github.com/fleetdm/fleet/issues/7456

> ```
> $ fleetctl generate mdm-apple --email=mozhu@gmail.com  --org-name="Acme, Inc."
> Sending certificate signing request (CSR) for Apple Push Notification service (APNs) to {mo@acme.com}... 
> Generating APNs key, Simple Certificate Enrollment Protocol (SCEP) certificate, and SCEP key...
> Error: Per Apple's terms of use, --email must be a work email (e.g. cannot be gmail, yahoo, hotmail)
> ```

it seems like we want to print `Sending ... Generating ...` even though there may have been an error. Does this mean that this is not considered a success message? 

Part of the issue that I think @roperzh is highlighting is that these messages are printed before we even attempt to do the actual work, and errors may occur at different stages. It is misleading to print `Sending certificate signing request ...` if we didn't even get to that stage because an error happened earlier than that.

Reply:
> I am not clear on what is considered "success messages"

@michalnicp the success message is this:

```
Success! 

Generated your APNs key at {/Users/Mo}/fleet-mdm-apple-apns.key

Generated your SCEP certificate at {/Users/Mo}/fleet-mdm-apple-scep-cert.crt

Generated your SCEP key at {/Users/Mo}/fleet-mdm-apple-scep-key.key

Go to your email to download a CSR from Fleet. Then, visit https://identity.apple.com/pushcert to upload the CSR. You should receive an APNs certificate in return from Apple. 

Next, use the generated certificates to deploy Fleet with `mdm` configuration: https://fleetdm.com/docs/deploying/configuration#mdm
```

Reply:
> we want to print `Sending ...` `Generating ...` even though there may have been an error. Does this mean that this is not considered a success message?

@michalnicp that's right. These are there to tell the users that Fleet is doing some thinking and working to make something happen (messages show exactly what)

Here's a Loom video walking through the expected success and error flows: https://www.loom.com/share/cb569d9adfe544939e35e5bbbdb191d0

> It is misleading to print Sending certificate signing request ... if we didn't even get to that stage because an error happened earlier than that.

I would argue that this isn't misleading (I could be wrong). For all the user knows, Fleet is trying to generate some things and send some things. If anything goes wrong under the hood, nothing is generated or sent and the user sees an error.